platform/lang-impl/src/com/intellij/execution/console/PrefixHistoryModel.kt

```diff
@@ -0,0 +1,289 @@
+    this.userContent = userContent
+  }
+
+  private var myEntries: ImmutableList<String>? = null
+  private var index: Int = 0
+
+
+  init {
+    resetIndex()
+  }
+
+  override fun resetEntries(entries: MutableList<String>) {
+    masterModel.resetEntries(entries)
+    resetIndex()
+  }
+
+  override fun addToHistory(statement: String?) {
+    if (statement.isNullOrEmpty()) {
+      return
+    }
+    masterModel.addToHistory(statement)
+    resetIndex()
+  }
+
+
+  override fun removeFromHistory(statement: String?) {
+    if (statement.isNullOrEmpty()) {
+      return
+    }
+    masterModel.removeFromHistory(statement)
+    resetIndex()
+  }
+
+  private fun resetIndex() {
+    myEntries = null
+    index = 0
+  }
+
+
+  override fun getHistoryNext(): TextWithOffset? {
+    val entries: ImmutableList<String> = myEntries ?: masterModel.entriesSnap.apply { index = this.size }
+    if (index <= 0) {
+      return null
+    }
+
+    val searchPrefix = getPrefixFn()
+    val indexOfLast = entries.subList(0, index).indexOfLast { it.startsWith(searchPrefix) || searchPrefix.isEmpty() }
+    if (indexOfLast == -1) {
+      return null
+    }
+    myEntries = entries
+    index = indexOfLast
+    return TextWithOffset(entries[index], searchPrefix.length)
+  }
+
+
+  override fun getHistoryPrev(): TextWithOffset? {
+    val entries: ImmutableList<String> = myEntries ?: return null
+    val searchPrefix = getPrefixFn()
+    val prevOffset = entries.subList(index + 1, entries.size).indexOfFirst { it.startsWith(searchPrefix) || searchPrefix.isEmpty() }
+    return if (prevOffset != -1) {
+      index += prevOffset + 1
+      TextWithOffset(entries[index], searchPrefix.length)
+    }
+    else if (userContent.startsWith(searchPrefix) || searchPrefix.isEmpty()) {
+      myEntries = null
+      index = 0
+      TextWithOffset(userContent, searchPrefix.length)
+
+    }
+    else {
+      null
+    }
+  }
+
+  //Can be remove when https://youtrack.jetbrains.com/issue/KT-19830 is fixed
+  override fun prevOnLastLine(): Boolean = true
+
+  override fun hasHistory(): Boolean = myEntries != null
+}
+
+private class MasterModel : ConsoleHistoryModel, EntriesWithPositionHolder() {
+  override fun setContent(userContent: String) = throw IllegalStateException("Should not be invoked")
+
+  val entriesSnap: ImmutableList<String>
+    get() = myState.entries
+
+
+  override fun resetEntries(ent: MutableList<String>) {
+    updateAtomically {
+      val newEntries = ent.toImmutableList()
+      return@updateAtomically State(newEntries, newEntries.size)
+    }
+  }
+
+
+  override fun addToHistory(statement: String?) {
+    val stmt = statement ?: return
+
+    updateAtomically { (entries) ->
+      val maxHistorySize = maxHistorySize
+      var newEntries = entries - stmt
+      if (newEntries.size >= maxHistorySize) {
+        newEntries = newEntries.removeAt(0)
+      }
+      newEntries += stmt
+
+      return@updateAtomically State(newEntries, newEntries.size)
```

Comment:
Minor: here and in some other places you can remove `return`, since it's a last statement in a lambda and it will be implicitly returned.